### PR TITLE
CSV Data Import Handler.

### PR DESCRIPTION
**Pull request by skmshaffer**
Writes feature to import CSV formatted catalog directly into the system. 

To test: 
- Ensure system is running in "dev" profile. 
- Add the --importcsvfile="/file/path/to/csv/file" argument to the system (either in IDE or as a command line argument while running). 
- Run the system and watch logs to verify file imported successfully.

### CODE REVIEW

File: src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+import org.apache.commons.csv.CSVRecord;
+import school.raikes.library.libraryserver.model.entity.*;
+
+@Builder
+public class CatalogCsvReader {
+
+  public static final int LONG_NAME_DETERMINANT = 6;
+  public static final Set<String> EMPTY_VALUES = ImmutableSet.of("", "-", "undefined");
+  public static final String DATE_FORMAT = "MM/dd/yyyy";
+  public static final String MANUAL_TAG_NAME = "MANUAL";
```
Comment (mathvav) [inline src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java:26]: Less nitpicky than average but still nitpicky: it took me about a minute to figure out what "manual tag" meant—and I was the one who specifically requested that feature lol. Can we rename it something more descriptive, such as `MANUALLY-ENTERED-METADATA`?

File: src/library-server/src/main/java/school/raikes/library/libraryserver/startup/CsvImportApplicationRunner.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+import school.raikes.library.libraryserver.loaders.CatalogCsvLoader;
+
+@Component
+@Profile("dev")
+@Slf4j
+public class CsvImportApplicationRunner implements ApplicationRunner {
+
+  private static final String CSV_FILE_PARAMETER = "importcsvfile";
```
[the remark this answers is not in the document]
Comment (mathvav) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/startup/CsvImportApplicationRunner.java:18]: Additionally, the CSV columns should be stored somewhere, with which ones are optional (probably just `MANUAL`). I can help with this if you'd like.

File: sql/db_init.sql [hunk 3]
```diff
@@ -17,18 +17,47 @@ drop table if exists tag;
 drop table if exists type;
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply sql/db_init.sql:18]: Definitely think those are good adds. It's an easy fix as those are fairly simply constraints. I'll walk through all of the entities and update where I see the need for additional constraints.

File: src/library-server/src/main/java/school/raikes/library/libraryserver/engine/IAuthorEngine.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package school.raikes.library.libraryserver.engine;
+
+import java.util.Map;
+import school.raikes.library.libraryserver.model.entity.Author;
+
+/** Engine interface for managing {@link Author}s. */
+public interface IAuthorEngine {
+  Iterable<Author> findAll();
+
+  Map<String, Author> loadNameAuthorMap();
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/engine/IAuthorEngine.java:10]: *grumbles and adds JavaDoc comments*

[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/engine/IAuthorEngine.java:10]: Done and updated.

File: src/library-server/src/main/java/school/raikes/library/libraryserver/config/CsvReaderConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.context.annotation.Configuration;
+import school.raikes.library.libraryserver.engine.IAuthorEngine;
+import school.raikes.library.libraryserver.engine.IBookEngine;
+import school.raikes.library.libraryserver.engine.IShelfEngine;
+import school.raikes.library.libraryserver.engine.ITagEngine;
+import school.raikes.library.libraryserver.readers.CatalogCsvReader;
+
+@Configuration
+public class CsvReaderConfiguration {
+
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/config/CsvReaderConfiguration.java:14]: Done

File: src/library-server/src/main/java/school/raikes/library/libraryserver/loaders/CatalogCsvLoader.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import school.raikes.library.libraryserver.engine.IBookEngine;
+import school.raikes.library.libraryserver.engine.IShelfEngine;
+import school.raikes.library.libraryserver.engine.ITagEngine;
+import school.raikes.library.libraryserver.readers.CatalogCsvReader;
+
+@Lazy
+@Component
+@Slf4j
+public class CatalogCsvLoader {
+
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/loaders/CatalogCsvLoader.java:22]: Done.

File: src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+  /* Maps used to track existing items so that new copies are properly linked to their respective book, author, etc */
+
+  @Builder.Default private final Map<Integer, Shelf> shelfMap = new HashMap<>();
+
+  @Builder.Default private final Map<String, Tag> tagMap = new HashMap<>();
+
+  @Builder.Default private final Map<String, Author> authorMap = new HashMap<>();
+
+  @Builder.Default private final Map<String, Book> isbnBookMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java:48]: It truly is brilliant and a wonderful addition to all Java projects. It saves you so much time.

File: src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+      String authors = row.get(AUTHORS_KEY).trim();
+      String publishDate = row.get(PUBLISH_DATE_KEY).trim();
+      String acquisitionDate = row.get(ACQUISITION_DATE_KEY).trim();
+      String shelf = row.get(SHELF_KEY).trim();
+      String manual = row.get(MANUAL_KEY).trim();
+
+      long parsedBarcode;
+      try {
+        // Assume Code 3 of 9 barcode, trim check digit and parse into value.
+        parsedBarcode = Long.parseLong(barcode.substring(0, barcode.length() - 1));
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java]: I built it out into a separate class. I try and grab it with the check digit, if the check digit isn't correct I just verify that it is a valid number (which is required by Code 39 barcodes).

File: src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        for (String author : Splitter.on(",").trimResults().split(authors)) {
+
+          // If just last names or organization (long) names just add and continue.
+          long nameLength = Splitter.on(CharMatcher.whitespace()).splitToStream(author).count();
+          if (nameLength == 1 || nameLength > LONG_NAME_DETERMINANT) {
+            parsedAuthors.add(new String[] {null, null, author});
+            continue;
+          }
+
+          HumanNameParserParser nameParser = new HumanNameParserParser(author.trim());
```
[the remark this answers is not in the document]
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java:121]: Me too, haha, especially when they're transpiled from PHP jsut for fun :P.

File: src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java (hunk 1 above)
Comment (skmshaffer) [reply src/library-server/src/main/java/school/raikes/library/libraryserver/readers/CatalogCsvReader.java:27]: Fixed!

**Comment by skmshaffer** [comment]
@mathvav Finished addressing all updates, ready for your final review!

**Pull request merged**